components/p3a/p3a_message.cc

```diff
@@ -67,6 +67,37 @@ void GenerateP3AMessage(uint64_t metric_hash,
   p3a_message->set_p3a_info(data, kDataLength);
 }
 
+base::Value GenerateP3AJsonMessage(uint64_t metric_hash,
+                                   uint64_t metric_value,
+                                   const MessageMetainfo& meta) {
+  base::Value result(base::Value::Type::DICTIONARY);
+
+  // Find out years of install and survey.
+  base::Time::Exploded exploded;
+  meta.date_of_survey.LocalExplode(&exploded);
+  DCHECK_GE(exploded.year, 999);
+  result.SetIntKey("yos", exploded.year);
+
+  meta.date_of_install.LocalExplode(&exploded);
+  DCHECK_GE(exploded.year, 999);
+  result.SetIntKey("yoi", exploded.year);
+
+  // Fill meta.
+  result.SetStringKey("country_code", meta.country_code);
+  result.SetStringKey("platform", meta.platform);
+  result.SetStringKey("version", meta.version);
+  result.SetStringKey("channel", meta.channel);
+  result.SetIntKey("woi", meta.woi);
+  result.SetIntKey("wos", meta.wos);
+  result.SetStringKey("refcode", meta.refcode);
+
+  // Set the metric
+  result.SetStringKey("metric_hash", base::NumberToString(metric_hash));
```

Comment:
yes, we can probably discuss it with @porteron

Reply:
That's a great suggestion @rillian . Hiding the packet length can be moved to a follow-up issue since we're not doing it in the existing system either.

Reply:
FWIW, @porteron said he's fine with submitting the metric name instead of the hash.